spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Regularizer.scala

```diff
@@ -0,0 +1,170 @@
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+
+import scala.reflect.ClassTag
+
+/**
+ * It is a trait for all regularizers.
+ * Any regularizers need to inherit the result.
+ *
+ * @tparam T type parameters [[Float]] or [[Double]]
+ */
+trait Regularizer[T]
+  extends Serializable {
+  var isRegualrized: Boolean = true
+
+  /**
+   * The method need to be override by the concrete regularizer class
+   * It accumulates the gradient of the regularization of `parameter` to `gradParameter`
+   *
+   * @param parameter the parameter that is regularized
+   * @param gradParameter the gradient of the parameter
+   */
+  def accRegularization(
+    parameter: Tensor[T],
+    gradParameter: Tensor[T]
+  ): Unit
+
+  /**
+   * Check the regularization is applied or not
+   *
+   * @param parameter the parameter that is regularized
+   * @param gradParameter the gradient of the parameter
+   * @return a boolean, if true, accumulates the gradient of regularization,
+   *         otherwise not.
+   */
+  protected def preCheck(
+    parameter: Tensor[T],
+    gradParameter: Tensor[T]
+  ): Boolean = {
+    if (null == parameter
+      || null == gradParameter
+      || !isRegualrized) {
+      false
+    } else {
+      true
+    }
+  }
+}
+
+/**
+ * Apply both L1 and L2 regularization
+ * @param l1 l1 regularization rate
+ * @param l2 l2 regularization rate
+ * @tparam T type parameters [[Float]] or [[Double]]
+ */
+@SerialVersionUID(- 5617491971070914067L)
+class L1L2Regularizer[T: ClassTag](
+  l1: Double,
+  l2: Double
+)(implicit ev: TensorNumeric[T])
+  extends Regularizer[T] {
+  override def accRegularization(
+    parameter: Tensor[T],
+    gradParameter: Tensor[T]
+  ): Unit = {
+    if (!preCheck(parameter, gradParameter)) return
+    accL1L2Regularization(l1, l2, parameter, gradParameter)
+  }
+
+  /**
+   * Accumulates the gradient of the l1, l2 regularization of `parameter`
+   * to `gradParameter`
+   *
+   * @param l1Alpha l1 regularization rate
+   * @param l2Alpha l2 regularization rate
+   * @param parameter the parameter that is regularized
+   * @param gradParameter the gradient of the parameter
+   */
+  private def accL1L2Regularization(
+    l1Alpha: Double,
+    l2Alpha: Double,
+    parameter: Tensor[T],
+    gradParameter: Tensor[T]
+  ): Unit = {
+    accL1Regularization(l1Alpha, parameter, gradParameter)
+    accL2Regularization(l2Alpha, parameter, gradParameter)
+  }
+
+  /**
+   * Accumulates the gradient of the l1 regularization of `parameter`
+   * to `gradParameter`
+   *
+   * @param alpha l1 regularization rate
+   * @param parameter the parameter that is regularized
+   * @param gradParameter the gradient of the parameter
+   */
+  private def accL1Regularization(
+    alpha: Double,
+    parameter: Tensor[T],
+    gradParameter: Tensor[T]
+  ): Unit = {
+    if (alpha != 0) gradParameter.add(ev.fromType(alpha),
+      l1SignBuffer.resizeAs(parameter).copy(parameter).sign())
```

Comment:
Do you mean something like 
"""
    if (alpha != 0) {
      gradParameter.map(parameter, (g, p) => {
        ev.plus(g, if (ev.isGreater(p, ev.zero)) ev.one else ev.negative(ev.one))
      })
    }
"""
that will a bit slower but without additional buffer, which leads less memory. 
which one do you think better?